api.go

```diff
@@ -177,6 +177,68 @@ func (r *lockedRandSource) Seed(seed int64) {
 	r.lk.Unlock()
 }
 
+// Redirect requests by re signing the request.
+func (c *Client) redirectHeaders(req *http.Request, via []*http.Request) error {
+	if len(via) >= 5 {
+		return errors.New("stopped after 5 redirects")
+	}
+	if len(via) == 0 {
+		return nil
+	}
+	lastRequest := via[len(via)-1]
+	var reAuth bool
+	for attr, val := range lastRequest.Header {
+		// if hosts do not match do not copy Authorization header
+		if attr == "Authorization" && req.Host != lastRequest.Host {
```

Comment:
I guess not only Host can change, even Path can do. A proxy can decide to tweak a little the url, like /user1/bucket/object to /bucket/object.

What about re-authenticating in all cases without condition, since this is a redirection.

Reply:
That is not expected to be handled here. Redirect is only to ensure that we are reAuth the host, changing paths is not supported. The server side redirection is supposed to redirect to a different hostname only.

Reply:
> The server side redirection is supposed to redirect to a different hostname only.

Are you saying this is an S3 requirement ?

Reply:
The specific requirement here is a customer, who wishes the client to be namespace aware. Namespace is cheaply provided by servers redirecting to servers where the incoming request belongs.  There is no mangling of path expected. Such scenario we should fail and it fails properly.

I could additionally add such a check along the lines of Host , but that is not expected as a requirement.